### PR TITLE
copr: the stats for normal column should be based on collation key

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: for https://github.com/pingcap/tidb/issues/14689 

Problem Summary:

We should use the collation key instead of its original value for a normal column to generate its statistics.

### What is changed and how it works?

What's Changed:

Use the collation sort key as the sampling data.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test https://github.com/pingcap/tidb/pull/18104

Side effects

- Performance regression
    - Consumes more CPU

### Release note 
- Fix the estimation error for a non-index column with collation enabled.

### CODE REVIEW

File: src/coprocessor/statistics/analyze.rs [hunk 1]
```diff
@@ -220,7 +228,31 @@ impl<S: Snapshot> SampleBuilder<S> {
                     pk_builder.append(&v);
                 }
             }
-            for (collector, val) in collectors.iter_mut().zip(cols_iter) {
+            for (i, (collector, val)) in collectors.iter_mut().zip(cols_iter).enumerate() {
+                if self.cols_info[i].get_tp() == FieldTypeTp::TinyBlob as i32
```
Comment (breeswish) [inline src/coprocessor/statistics/analyze.rs:232]: Maybe you can use `self.cols_info[i].as_accessor().is_string_like()`

File: src/coprocessor/statistics/analyze.rs [hunk 2]
```diff
@@ -220,7 +228,31 @@ impl<S: Snapshot> SampleBuilder<S> {
...
+                    || self.cols_info[i].get_tp() == FieldTypeTp::MediumBlob as i32
+                    || self.cols_info[i].get_tp() == FieldTypeTp::Blob as i32
+                    || self.cols_info[i].get_tp() == FieldTypeTp::VarChar as i32
+                    || self.cols_info[i].get_tp() == FieldTypeTp::VarString as i32
+                    || self.cols_info[i].get_tp() == FieldTypeTp::String as i32
+                {
+                    let sorted_val = match_template_collator! {
+                        TT, match Collation::from_i32(self.cols_info[i].get_collation())? {
```
Comment (breeswish) [inline src/coprocessor/statistics/analyze.rs:240]: `match self.cols_info[i].as_accessor().collation()`

File: src/coprocessor/statistics/analyze.rs [hunk 3]
```diff
@@ -220,7 +228,31 @@ impl<S: Snapshot> SampleBuilder<S> {
...
+                            Collation::TT => {
+                                let mut mut_val = val.as_slice();
+                                let decoded_val = table::decode_col_value(&mut mut_val, &mut EvalContext::default(), &self.cols_info[i])?;
+                                if decoded_val == datum::Datum::Null {
```
Comment (breeswish) [inline src/coprocessor/statistics/analyze.rs:244]: prefer not to use functions in `datum::`

File: src/coprocessor/statistics/analyze.rs [hunk 4]
```diff
@@ -220,7 +228,31 @@ impl<S: Snapshot> SampleBuilder<S> {
...
+                    || self.cols_info[i].get_tp() == FieldTypeTp::Blob as i32
+                    || self.cols_info[i].get_tp() == FieldTypeTp::VarChar as i32
+                    || self.cols_info[i].get_tp() == FieldTypeTp::VarString as i32
+                    || self.cols_info[i].get_tp() == FieldTypeTp::String as i32
+                {
+                    let sorted_val = match_template_collator! {
+                        TT, match Collation::from_i32(self.cols_info[i].get_collation())? {
+                            Collation::TT => {
+                                let mut mut_val = val.as_slice();
+                                let decoded_val = table::decode_col_value(&mut mut_val, &mut EvalContext::default(), &self.cols_info[i])?;
```
Comment (breeswish) [inline src/coprocessor/statistics/analyze.rs:243]: prefer not to use functions in `table::`

@iosmanthus Please give some advice for better alternatives

**Comment by breeswish** [comment]
@Renkai PTAL

**Comment by winoros** [comment]
i'll push the new commit after another review

File: src/coprocessor/statistics/analyze.rs (hunk 1 above)
Comment (breeswish) [reply src/coprocessor/statistics/analyze.rs:232]: @winoros `as_accessor` is available for both `ColumnInfo` and `FieldType`, after `use xxxx::FieldTypeAccessor`

File: src/coprocessor/statistics/analyze.rs (hunk 3 above)
Comment (winoros) [reply src/coprocessor/statistics/analyze.rs]: I found that i may misread your comment. You mean don't use anything in `datum` package?
I may need more suggestions to move it. I use the same way in TiDB :joy:

File: src/coprocessor/statistics/analyze.rs [hunk 5]
```diff
@@ -176,9 +180,12 @@ impl<S: Snapshot> SampleBuilder<S> {
         }
 
         let mut col_len = cols_info.len();
-        if cols_info[0].get_pk_handle() {
+        let vec_cols_info = if cols_info[0].get_pk_handle() {
             col_len -= 1;
-        }
+            cols_info.to_vec()[1..col_len + 1].to_vec()
```
Comment (iosmanthus) [inline src/coprocessor/statistics/analyze.rs:185]: `cols_info[1..].to_vec()` instead?

**Comment by winoros** [comment]
This pr should be merged together with tidb's.

**Pull request merged**